File: insertSort_dichotomy.py
def insertSort_dichotomy(a):
    n = len(a)
    for i in range(1,n):
        insert = a[i]
        left = 0
        right = i-1
        while left <= right:
            mid = (left + right)/2
            if a[mid]<insert:
                left = mid + 1
            else:
                right = mid - 1

        for j in range(i-1, left-1, -1):
            a[j+1] = a[j]
        a[left] = insert

A = [6, 5, 3, 1, 8, 7, 2, 4]
print (A)
insertSort_dichotomy(A)
print (A)

File: sort_cpp/insertSort_dichotomy.cpp
#include <iostream>
using namespace std;

void insertSort_dichotomy(int *A, int n){
  for (int i=1;i<n;i++){
    int left = 0, right=i;
    int mid = (left+right)/2;
    int target = A[i];
    while (left<=right){
      if (A[i] < A[mid])
        right = mid - 1;
      else
        left = mid + 1;
    }
    for (int j=right;j<=left;j--){
      A[j+1] = A[j];
    }
    A[left] = target;
  }
};

int main()
{
  int A[10] = {1,2,9,4,5,10,7,6,8,3};
  int n = 10;
  for (int i=0; i<n; i++){
    cout << A[i] << endl;
  }

  insertSort_dichotomy(A, n);

  for (int i=0; i<n; i++){
    cout << A[i] << endl;
  }
}
